File: stock-app/src/utils/technicals.ts
export function sma(data: number[], period: number): number {
  if (data.length < period) return 0;
  const slice = data.slice(-period);
  return slice.reduce((a, b) => a + b, 0) / period;
}

export function emaArray(data: number[], period: number): number[] {
  if (data.length === 0) return [];
  const k = 2 / (period + 1);
  const result: number[] = [data[0]];
  for (let i = 1; i < data.length; i++) {
    result.push(data[i] * k + result[i - 1] * (1 - k));
  }
  return result;
}

export function rsi(closes: number[], period: number = 14): number {
  if (closes.length < period + 1) return 50;

  let gains = 0;
  let losses = 0;

  for (let i = 1; i <= period; i++) {
    const diff = closes[i] - closes[i - 1];
    if (diff > 0) gains += diff;
    else losses -= diff;
  }

  let avgGain = gains / period;
  let avgLoss = losses / period;

  for (let i = period + 1; i < closes.length; i++) {
    const diff = closes[i] - closes[i - 1];
    avgGain = (avgGain * (period - 1) + Math.max(0, diff)) / period;
    avgLoss = (avgLoss * (period - 1) + Math.max(0, -diff)) / period;
  }

  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export interface MacdResult {
  macdLine: number;
  signalLine: number;
  histogram: number;
  isBullish: boolean;
  crossedUp: boolean; // histogram turned positive this period
}

export function macd(
  closes: number[],
  fastPeriod = 12,
  slowPeriod = 26,
  signalPeriod = 9
): MacdResult {
  if (closes.length < slowPeriod + signalPeriod) {
    return { macdLine: 0, signalLine: 0, histogram: 0, isBullish: false, crossedUp: false };
  }

  const ema12 = emaArray(closes, fastPeriod);
  const ema26 = emaArray(closes, slowPeriod);
  const macdLine = ema12.map((v, i) => v - ema26[i]);
  const signalArr = emaArray(macdLine.slice(-(signalPeriod + 10)), signalPeriod);

  const lastMacd = macdLine[macdLine.length - 1];
  const prevMacd = macdLine[macdLine.length - 2] ?? 0;
  const lastSignal = signalArr[signalArr.length - 1];
  const prevSignal = signalArr[signalArr.length - 2] ?? 0;
  const histogram = lastMacd - lastSignal;
  const prevHistogram = prevMacd - prevSignal;

  return {
    macdLine: lastMacd,
    signalLine: lastSignal,
    histogram,
    isBullish: histogram > 0,
    crossedUp: histogram > 0 && prevHistogram <= 0,
  };
}

export function bollingerBands(
  closes: number[],
  period = 20,
  stdDevMultiplier = 2
): { upper: number; middle: number; lower: number } {
  if (closes.length < period) {
    const last = closes[closes.length - 1] ?? 0;
    return { upper: last, middle: last, lower: last };
  }
  const slice = closes.slice(-period);
  const mean = slice.reduce((a, b) => a + b, 0) / period;
  const variance = slice.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / period;
  const stdDev = Math.sqrt(variance);
  return {
    upper: mean + stdDev * stdDevMultiplier,
    middle: mean,
    lower: mean - stdDev * stdDevMultiplier,
  };
}

export function volumeRatio(volumes: number[], period = 20): number {
  if (volumes.length < period + 1) return 1;
  const avg = sma(volumes.slice(0, -1), period);
  if (avg === 0) return 1;
  return volumes[volumes.length - 1] / avg;
}

export function periodReturn(closes: number[], days: number): number {
  if (closes.length < days + 1) return 0;
  const current = closes[closes.length - 1];
  const past = closes[closes.length - 1 - days];
  if (past === 0) return 0;
  return (current - past) / past;
}

File: scripts/technicals.py
"""Technical indicators (mirrors stock-app/src/utils/technicals.ts)."""


def sma(data: list, period: int) -> float:
    if len(data) < period:
        return 0
    return sum(data[-period:]) / period


def ema_array(data: list, period: int) -> list:
    if not data:
        return []
    k = 2 / (period + 1)
    result = [data[0]]
    for v in data[1:]:
        result.append(v * k + result[-1] * (1 - k))
    return result


def rsi(closes: list, period: int = 14) -> float:
    if len(closes) < period + 1:
        return 50

    gains = 0.0
    losses = 0.0
    for i in range(1, period + 1):
        diff = closes[i] - closes[i - 1]
        if diff > 0:
            gains += diff
        else:
            losses -= diff

    avg_gain = gains / period
    avg_loss = losses / period

    for i in range(period + 1, len(closes)):
        diff = closes[i] - closes[i - 1]
        avg_gain = (avg_gain * (period - 1) + max(0, diff)) / period
        avg_loss = (avg_loss * (period - 1) + max(0, -diff)) / period

    if avg_loss == 0:
        return 100
    rs = avg_gain / avg_loss
    return 100 - 100 / (1 + rs)


def macd(closes: list, fast_period: int = 12, slow_period: int = 26, signal_period: int = 9) -> dict:
    if len(closes) < slow_period + signal_period:
        return {"macd_line": 0, "signal_line": 0, "histogram": 0, "is_bullish": False, "crossed_up": False}

    ema_fast = ema_array(closes, fast_period)
    ema_slow = ema_array(closes, slow_period)
    macd_line_arr = [f - s for f, s in zip(ema_fast, ema_slow)]
    signal_arr = ema_array(macd_line_arr[-(signal_period + 10):], signal_period)

    last_macd = macd_line_arr[-1]
    prev_macd = macd_line_arr[-2] if len(macd_line_arr) >= 2 else 0
    last_signal = signal_arr[-1]
    prev_signal = signal_arr[-2] if len(signal_arr) >= 2 else 0
    histogram = last_macd - last_signal
    prev_histogram = prev_macd - prev_signal

    return {
        "macd_line": last_macd,
        "signal_line": last_signal,
        "histogram": histogram,
        "is_bullish": histogram > 0,
        "crossed_up": histogram > 0 and prev_histogram <= 0,
    }


def volume_ratio(volumes: list, period: int = 20) -> float:
    if len(volumes) < period + 1:
        return 1
    avg = sma(volumes[:-1], period)
    if avg == 0:
        return 1
    return volumes[-1] / avg


def period_return(closes: list, days: int) -> float:
    if len(closes) < days + 1:
        return 0
    current = closes[-1]
    past = closes[-1 - days]
    if past == 0:
        return 0
    return (current - past) / past
